pkg/apis/experimental/types.go

```diff
@@ -639,3 +639,36 @@ type ClusterAutoscalerList struct {
 
 	Items []ClusterAutoscaler `json:"items"`
 }
+
+// Selector is a label query over a set of API objects. MatchLabels and
+// MatchExpressions are ANDed. An empty Selector matches all objects. A
+// null Selector matches no objects.
+type Selector struct {
+	// MatchLabels is a list of labels.
+	MatchLabels map[string]string `json:"matchLabels,omitempty"`
+	// MatchExpressions is a list of requirement selectors.
+	MatchExpressions []SelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// SelectorRequirement is a selector that contains values, a key
+// and an operator that relates the key and values.
+type SelectorRequirement struct {
+	// Key is the label key that the selector applies to.
+	Key string `json:"key"`
+	// Operator represents a key's relationship
+	// to a set of values.
+	Operator SelectorOperator `json:"operator"`
+	// StringValues is a set of string values. If the operator is In or NotIn,
+	// the values set must be non-empty.
+	StringValues []string `json:"stringValues,omitempty"`
+}
+
+// SelectorOperator is the set of operators that can be used in
+// a selector requirement.
+type SelectorOperator string
+
+const (
+	SelectorOpIn     SelectorOperator = "In"
+	SelectorOpNotIn  SelectorOperator = "NotIn"
+	SelectorOpExists SelectorOperator = "Exists"
```

Comment:
As always, these decisions are tradeoffs, not black-and-white. Your first two bullets boil down to performance, while your second two bullets boil down to understandability. These are great properties to have. But IMO here they need to be balanced against another dimension of usability, which is expressiveness.

With respect to the expressability/performance tradeoff (your first two bullets), I think it would be fine for us to say that simple selectors will be evaluated fast, while complex selectors will be evaluated slowly. With respect to the third, I think understandability means different things to different people -- you could imagine that in the absence of regexp. people might have to create horribly convoluted labels and selectors in order to achieve their goal -- i.e. although regexp's are notoriously hard for humans to parse, they are much more compact than alternative representations. WRT the performance impact of regexp's, again we could say that if your selector uses a regexp it's going to be slow.

Here are my comments on the specific features under discussion in this section:

* Does Not Exist: It seems to me that "does not exist" is justifiable on the grounds that for any kind of scenario where you'd use "exists" to target some types of pods to certain types of nodes, you may want to tell the other types of pods to avoid those nodes. 

* Numeric comparison: Even if we come up with a way to easily introduce new resource types (something we haven't spec'd out yet, which makes me very uncomfortable using it as a justification for not supporting numeric comparison here), that will presumably only be available to cluster admins and only makes sense for nodes. It does not help users who want to tag their own objects (for example, "pod is from a release > version 6").  

* Regexp's: I actually don't feel super-strongly one way or the other on this one. However, note that Marathon supports it (see LIKE/UNLIKE here: https://github.com/mesosphere/marathon/blob/master/docs/docs/constraints.md )

On "escape hatches": I think escape hatches are orthogonal and shouldn't be used to justify overly restricting the syntax for selectors. They have a lot of drawbacks like race conditions and extra overhead for the user to actually build and deploy the thing (whether it's an agent that adds labels, or one that evaluates selectors).